### PR TITLE
New loader: Update test_loader.py

### LINKED ISSUE #718: Update test_loader.py
`test_loader.py` needs to be updated to include:

- [x] participant_id column in `test_bids_df_anat`
- [x] participant_id and sample_id in `test_bids_df_microscopy_png`
- [x] a test for multiple folders in path_data
- [x] a test for when `roi_suffix` is not None
- [x] a test for CT-scan data

This should be done after updating the data-testing repo accordingly (https://github.com/ivadomed/data-testing/issues/5#issuecomment-799567954)

### PR DESCRIPTION
**Pull request by mariehbourget**
#### GitHub

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [x] I've linked relevant issues in the PR body
- [x] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [x] I've assigned a reviewer

#### PR contents

- [x] I've consulted [ivadomed's internal developer documentation](https://github.com/ivadomed/ivadomed/wiki) to ensure my contribution is in line with any relevant design decisions
- [x] I've added [relevant tests](https://github.com/ivadomed/ivadomed/wiki/tests) for my contribution
- [x] I've updated the [relevant documentation](https://github.com/ivadomed/ivadomed/wiki/documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Description
This PR updates `test_loader.py` following the integration of the new loader (#699).
It also includes new tests for multiple folders in path_data, and for ct-scan file indexing.

The tests won't pass until PR #[13](https://github.com/ivadomed/data-testing/pull/13) on data-testing repo is merged and released, and the link updated in download_data [here](https://github.com/ivadomed/ivadomed/blob/5564e6345a0bffdee52631cae4f9d6f0fe50c122/ivadomed/scripts/download_data.py#L25).

## Linked issues
Fixes #718

### CODE REVIEW

File: ivadomed/loader/utils.py [hunk 1]
```diff
@@ -737,7 +737,7 @@ def create_bids_dataframe(self):
             df_next = layout.to_df(metadata=True)
 
             # Add filename column
-            df_next['filename'] = df_next['path'].apply(os.path.basename)
+            df_next.insert(1,'filename', df_next['path'].apply(os.path.basename))
```
Comment (jcohenadad) [inline ivadomed/loader/utils.py:740]:
```suggestion
            df_next.insert(1, 'filename', df_next['path'].apply(os.path.basename))
```

Comment (mariehbourget) [reply ivadomed/loader/utils.py]: Fixed in 5b1f77c

**Comment by mariehbourget** [comment]
This PR is ready for review.
A new version of data-testing has been [released](https://github.com/ivadomed/data-testing/releases/tag/r20210322) to match the new tests.
To test, run `ivadomed_download_data -d data_testing -o data_testing`, then `pytest testing/unit_tests/`

**Comment by charleygros** [comment]
> To test, run `ivadomed_download_data -d data_testing -o data_testing`, then `pytest testing/unit_tests/`

I run the command and got the below error. Any ideas? maybe an installation issue on my hand?

image

**Comment by mariehbourget** [comment]
> Any ideas?

`csv-diff` is a dependency that I added to `requirements.txt` in PR #584 with the first test for the loader. Can you try to re-install the requirements in your environment?

**Review by gisellemartel — APPROVED** [review]
Tested locally and it works! Note that I did need to reinstall the pip dependencies prior to running pytest

**Comment by charleygros** [comment]
I re-installed as suggested and pytests are now running. I still get these errors tho:

```
(venv-ivadomed) chgroc@rosenberg:~/ivadomed$ pytest testing/unit_tests/
================================================= test session starts ==================================================platform linux -- Python 3.6.9, pytest-6.1.2, py-1.9.0, pluggy-0.13.1
rootdir: /home/GRAMES.POLYMTL.CA/chgroc/ivadomed
plugins: ordering-0.6, console-scripts-1.1.0
collected 175 items

testing/unit_tests/test_HeMIS.py ..                                                                              [  1%]
testing/unit_tests/test_adaptative.py .                                                                          [  1%]
testing/unit_tests/test_automate_training.py ....                                                                [  4%]
testing/unit_tests/test_bounding_box.py ....                                                                     [  6%]
testing/unit_tests/test_loader.py FF.....                                                                        [ 10%]
testing/unit_tests/test_losses.py ...........................................                                    [ 34%]
testing/unit_tests/test_metrics.py .......................                                                       [ 48%]
testing/unit_tests/test_mixup.py ....                                                                            [ 50%]
testing/unit_tests/test_model.py ......                                                                          [ 53%]
testing/unit_tests/test_onnx.py .                                                                                [ 54%]
testing/unit_tests/test_orientation.py .                                                                         [ 54%]
testing/unit_tests/test_postprocessing.py ........                                                               [ 59%]
testing/unit_tests/test_rgb.py .                                                                                 [ 60%]
testing/unit_tests/test_sampler.py .                                                                             [ 60%]
testing/unit_tests/test_segment_volume.py F.                                                                     [ 61%]
testing/unit_tests/test_slice_filter.py ........                                                                 [ 66%]
testing/unit_tests/test_split_dataset.py ......                                                                  [ 69%]
testing/unit_tests/test_tensorboard_save.py .                                                                    [ 70%]
testing/unit_tests/test_testing.py ..                                                                            [ 71%]
testing/unit_tests/test_training_time.py ......                                                                  [ 74%]
testing/unit_tests/test_transfer_learning.py ...                                                                 [ 76%]
testing/unit_tests/test_transforms.py .........................................                                  [100%]

======================================================= FAILURES =======================================================___________________________________ test_bids_df_microscopy_png[loader_parameters0] ____________________________________
loader_parameters = {'bids_config': 'ivadomed/config/config_bids.json', 'contrast_params': {'contrast_lst': []}, 'extensions': ['.png'], 'path_data': ['tmp/data_testing/microscopy_png'], ...}

    @pytest.mark.parametrize('loader_parameters', [{
        "path_data": [os.path.join(__data_testing_dir__, "microscopy_png")],
        "bids_config": "ivadomed/config/config_bids.json",
        "target_suffix": [["_seg-myelin-manual", "_seg-axon-manual"]],
        "extensions": [".png"],
        "roi_params": {"suffix": None, "slice_filter_roi": None},
        "contrast_params": {"contrast_lst": []}
        }])
    def test_bids_df_microscopy_png(loader_parameters):
        """
        Test for microscopy png file format
        Test for _sessions.tsv and _scans.tsv files
        Test for target_suffix as a nested list
        Test for when no contrast_params are provided
        """

        bids_df = imed_loader_utils.BidsDataframe(loader_parameters, __tmp_dir__, derivatives=True)
        df_test = bids_df.df.drop(columns=['path'])
        # TODO: modify df_ref.csv file in data-testing dataset to include "participant_id"
        # and "sample_id" columns, then delete next line
        df_test = df_test.drop(columns=['participant_id', 'sample_id'])
        df_test = df_test.sort_values(by=['filename']).reset_index(drop=True)
        csv_ref = os.path.join(loader_parameters["path_data"][0], "df_ref.csv")
        csv_test = os.path.join(loader_parameters["path_data"][0], "df_test.csv")
        df_test.to_csv(csv_test, index=False)
        diff = csv_diff.compare(csv_diff.load_csv(open(csv_ref)), csv_diff.load_csv(open(csv_test)))
>       assert diff == {'added': [], 'removed': [], 'changed': [], 'columns_added': [], 'columns_removed': []}
E       AssertionError: assert {'added': [{'...ant_id'], ...} == {'added': [],...ved': [], ...}
E         Omitting 2 identical items, use -vv to show
E         Differing items:
E         {'columns_removed': ['sample_id', 'participant_id']} != {'columns_removed': []}
E         {'added': [{'BitDepth': '8', 'BodyPart': 'CSPINE', 'Channels': '1', 'Environment': 'exvivo', ...}, {'BitDepth': '', 'B...Channels': '', 'Environment': '', ...}, {'BitDepth': '', 'BodyPart': '', 'Channels': '', 'Environment': '', ...}, ...]} != {'added': []}
E         {'removed': [{'BitDepth': '8', 'BodyPart': 'CSPINE', 'Channels': '1', 'Environment': 'exvivo', ...}, {'BitDepth': '', ...Channels': '', 'Environment': '', ...}, {'BitDepth': '', 'Bo...
E
E         ...Full output truncated (2 lines hidden), use '-vv' to show

testing/unit_tests/test_loader.py:42: AssertionError
------------------------------------------------- Captured stdout call -------------------------------------------------Dataframe has been saved in tmp/bids_dataframe.csv.
________________________________________ test_bids_df_anat[loader_parameters0] _________________________________________
loader_parameters = {'contrast_params': {'contrast_lst': ['T1w', 'T2w']}, 'extensions': [], 'path_data': ['tmp/data_testing'], 'roi_params': {'slice_filter_roi': None, 'suffix': None}, ...}

    @pytest.mark.parametrize('loader_parameters', [{
        "path_data": [__data_testing_dir__],
        "target_suffix": ["_seg-manual"],
        "extensions": [],
        "roi_params": {"suffix": None, "slice_filter_roi": None},
        "contrast_params": {"contrast_lst": ["T1w", "T2w"]}
        }])
    def test_bids_df_anat(loader_parameters):
        """
        Test for MRI anat nii.gz file format
        Test for when no file extensions are provided
        Test for multiple target_suffix
        TODO: modify test and "df_ref.csv" file in data-testing dataset to test behavior when "roi_suffix" is not None
        """

        bids_df = imed_loader_utils.BidsDataframe(loader_parameters, __tmp_dir__, derivatives = True)
        df_test = bids_df.df.drop(columns=['path'])
        # TODO: modify df_ref.csv file in data-testing dataset to include "participant_id"
        # column then delete next line
        df_test = df_test.drop(columns=['participant_id'])
        df_test = df_test.sort_values(by=['filename']).reset_index(drop=True)
        csv_ref = os.path.join(loader_parameters["path_data"][0], "df_ref.csv")
        csv_test = os.path.join(loader_parameters["path_data"][0], "df_test.csv")
        df_test.to_csv(csv_test, index=False)
        diff = csv_diff.compare(csv_diff.load_csv(open(csv_ref)), csv_diff.load_csv(open(csv_test)))
>       assert diff == {'added': [], 'removed': [], 'changed': [],
                        'columns_added': [], 'columns_removed': []}
E       AssertionError: assert {'added': [{'...ant_id'], ...} == {'added': [],...ved': [], ...}
E         Omitting 2 identical items, use -vv to show
E         Differing items:
E         {'columns_removed': ['participant_id']} != {'columns_removed': []}
E         {'added': [{'AcquisitionMatrixPE': '260', 'AcquisitionNumber': '1', 'AcquisitionTime': '12:33:7.032500', 'BaseResoluti...: '320', ...}, {'AcquisitionMatrixPE': '', 'AcquisitionNumber': '', 'AcquisitionTime': '', 'BaseResolution': '', ...}]} != {'added': []}
E         {'removed': [{'AcquisitionMatrixPE': '260', 'AcquisitionNumber': '1', 'AcquisitionTime': '12:33:7.032500', 'BaseResolu...on': '', ...}, {'AcquisitionMatrixPE': '', 'AcquisitionNumber': '', 'Acqu...
E
E         ...Full output truncated (2 lines hidden), use '-vv' to show

testing/unit_tests/test_loader.py:70: AssertionError
------------------------------------------------- Captured stdout call -------------------------------------------------Dataframe has been saved in tmp/bids_dataframe.csv.
-------------------------------------------------- Captured log call ---------------------------------------------------WARNING  ivadomed.loader.utils:utils.py:872 Missing derivatives for sub-unf02_T1w.nii.gz. Skipping.
WARNING  ivadomed.loader.utils:utils.py:872 Missing derivatives for sub-unf02_T2w.nii.gz. Skipping.
WARNING  ivadomed.loader.utils:utils.py:872 Missing derivatives for sub-unf03_T1w.nii.gz. Skipping.
WARNING  ivadomed.loader.utils:utils.py:872 Missing derivatives for sub-unf03_T2w.nii.gz. Skipping.
________________________________________________ test_segment_volume_2d ________________________________________________
    def test_segment_volume_2d():
        model = imed_models.Unet(in_channel=1,
                                 out_channel=1,
                                 depth=2,
                                 drop_rate=DROPOUT,
                                 bn_momentum=BN)

        if not os.path.exists(PATH_MODEL):
            os.mkdir(PATH_MODEL)

        torch.save(model, os.path.join(PATH_MODEL, "model_test.pt"))
        config = {
            "loader_parameters": {
                "slice_filter_params": {
                    "filter_empty_mask": False,
                    "filter_empty_input": False
                },
                "roi_params": {
                    "suffix": None,
                    "slice_filter_roi": 10
                },
                "slice_axis": "axial"
            },
            "transformation": {
                "Resample": {"wspace": 0.75, "hspace": 0.75},
                "ROICrop": {"size": [48, 48]},
                "RandomTranslation": {
                    "translate": [0.03, 0.03],
                    "applied_to": ["im", "gt"],
                    "dataset_type": ["training"]
                },
                "NumpyToTensor": {},
                "NormalizeInstance": {"applied_to": ["im"]}
            },
            "postprocessing": {},
            "training_parameters": {
                "batch_size": BATCH_SIZE
            }
        }

        PATH_CONFIG = os.path.join(PATH_MODEL, 'model_test.json')
        with open(PATH_CONFIG, 'w') as fp:
            json.dump(config, fp)

>       nib_lst, _ = imed_inference.segment_volume(PATH_MODEL, [IMAGE_PATH], ROI_PATH)

testing/unit_tests/test_segment_volume.py:72:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

folder_model = 'tmp/model_test', fname_images = ['tmp/data_testing/sub-unf01/anat/sub-unf01_T1w.nii.gz']
gpu_id = 'tmp/data_testing/derivatives/labels/sub-unf01/anat/sub-unf01_T1w_seg-manual.nii.gz', options = None

    def segment_volume(folder_model, fname_images, gpu_id=0, options=None):
        """Segment an image.

        Segment an image (`fname_image`) using a pre-trained model (`folder_model`). If provided, a region of interest
        (`fname_roi`) is used to crop the image prior to segment it.

        Args:
            folder_model (str): foldername which contains
                (1) the model ('folder_model/folder_model.pt') to use
                (2) its configuration file ('folder_model/folder_model.json') used for the training,
                see https://github.com/neuropoly/ivadomed/wiki/configuration-file
            fname_images (list): list of image filenames (e.g. .nii.gz) to segment. Multichannel models require multiple                images to segment, e.i., len(fname_images) > 1.
            gpu_id (int): Number representing gpu number if available.
            options (dict): Contains postprocessing steps and prior filename (fname_prior) which is an image filename
                (e.g., .nii.gz) containing processing information (e.i., spinal cord segmentation, spinal location or MS                lesion classification)
                e.g., spinal cord centerline, used to crop the image prior to segment it if provided.
                The segmentation is not performed on the slices that are empty in this image.

        Returns:
            list: List of nibabel objects containing the soft segmentation(s), one per prediction class.
            list: List of target suffix associated with each prediction in `pred_list`

        """
        # Define device
        cuda_available = torch.cuda.is_available()
>       device = torch.device("cpu") if not cuda_available else torch.device("cuda:" + str(gpu_id))
E       RuntimeError: Could not parse device index 'tmp/data_testing/derivatives/labels/sub-unf01/anat/sub-unf01_T1w_seg-manual.nii.gz' in device string 'cuda:tmp/data_testing/derivatives/labels/sub-unf01/anat/sub-unf01_T1w_seg-manual.nii.gz'

ivadomed/inference.py:131: RuntimeError
=================================================== warnings summary ===================================================../venv-ivadomed/lib/python3.6/distutils/__init__.py:4
  /home/GRAMES.POLYMTL.CA/chgroc/venv-ivadomed/lib/python3.6/distutils/__init__.py:4: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp

testing/unit_tests/test_HeMIS.py: 13 warnings
testing/unit_tests/test_adaptative.py: 13 warnings
testing/unit_tests/test_bounding_box.py: 26 warnings
testing/unit_tests/test_loader.py: 24 warnings
testing/unit_tests/test_orientation.py: 13 warnings
testing/unit_tests/test_sampler.py: 13 warnings
testing/unit_tests/test_slice_filter.py: 78 warnings
testing/unit_tests/test_testing.py: 26 warnings
testing/unit_tests/test_training_time.py: 78 warnings
  /home/GRAMES.POLYMTL.CA/chgroc/venv-ivadomed/lib/python3.6/site-packages/bids/layout/layout.py:526: DeprecationWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.
    data.loc[o] = pd.Series()

testing/unit_tests/test_onnx.py::test_onnx
  /home/GRAMES.POLYMTL.CA/chgroc/venv-ivadomed/lib/python3.6/site-packages/torch/nn/functional.py:1901: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
    if size_prods == 1:

testing/unit_tests/test_onnx.py::test_onnx
  /home/GRAMES.POLYMTL.CA/chgroc/venv-ivadomed/lib/python3.6/site-packages/torch/nn/functional.py:2864: TracerWarning: torch.tensor results are registered as constants in the trace. You can safely ignore this warning if you use this function to create tensors out of constant variables that would be the same every time you call this function. In any other case, this might cause the trace to be incorrect.
    dtype=torch.float32)).float())) for i in range(dim)]

testing/unit_tests/test_training_time.py::test_unet_time[config5-target_lst0-train_lst0]
  /home/GRAMES.POLYMTL.CA/chgroc/venv-ivadomed/lib/python3.6/site-packages/numpy/core/fromnumeric.py:3373: RuntimeWarning: Mean of empty slice.
    out=out, **kwargs)

testing/unit_tests/test_training_time.py::test_unet_time[config5-target_lst0-train_lst0]
  /home/GRAMES.POLYMTL.CA/chgroc/venv-ivadomed/lib/python3.6/site-packages/numpy/core/_methods.py:170: RuntimeWarning: invalid value encountered in double_scalars
    ret = ret.dtype.type(ret / rcount)

testing/unit_tests/test_training_time.py::test_unet_time[config5-target_lst0-train_lst0]
  /home/GRAMES.POLYMTL.CA/chgroc/venv-ivadomed/lib/python3.6/site-packages/numpy/core/_methods.py:234: RuntimeWarning: Degrees of freedom <= 0 for slice
    keepdims=keepdims)

testing/unit_tests/test_training_time.py::test_unet_time[config5-target_lst0-train_lst0]
  /home/GRAMES.POLYMTL.CA/chgroc/venv-ivadomed/lib/python3.6/site-packages/numpy/core/_methods.py:195: RuntimeWarning: invalid value encountered in true_divide
    arrmean, rcount, out=arrmean, casting='unsafe', subok=False)

testing/unit_tests/test_training_time.py::test_unet_time[config5-target_lst0-train_lst0]
  /home/GRAMES.POLYMTL.CA/chgroc/venv-ivadomed/lib/python3.6/site-packages/numpy/core/_methods.py:226: RuntimeWarning: invalid value encountered in double_scalars
    ret = ret.dtype.type(ret / rcount)

testing/unit_tests/test_transforms.py::test_Clahe[clahe0-im_seg0]
  /home/GRAMES.POLYMTL.CA/chgroc/venv-ivadomed/lib/python3.6/site-packages/skimage/util/dtype.py:226: DeprecationWarning: Converting `np.inexact` or `np.floating` to a dtype is deprecated. The current result is `float64` which is not strictly correct.
    dtypeobj_out = np.dtype(dtype)

-- Docs: https://docs.pytest.org/en/stable/warnings.html
=============================================== short test summary info ================================================FAILED testing/unit_tests/test_loader.py::test_bids_df_microscopy_png[loader_parameters0] - AssertionError: assert {'...FAILED testing/unit_tests/test_loader.py::test_bids_df_anat[loader_parameters0] - AssertionError: assert {'added': [{...FAILED testing/unit_tests/test_segment_volume.py::test_segment_volume_2d - RuntimeError: Could not parse device index...====================================(
```

**Comment by mariehbourget** [comment]
>         # TODO: modify df_ref.csv file in data-testing dataset to include "participant_id"
>         # and "sample_id" columns, then delete next line

I don't know for the `test_segment_volume.py`, but in the `test_loader.py` case, the 2 lines above are from an old version. It seems you are not on the latest version + this branch.

**Pull request merged**